Reach skip nav example (#21376)
gatsbyjs/gatsby: examples/using-reach-skip-nav/README.md

> # Using @reach/skip-nav

Reviewer: I still feel like this example needs to describe what it does, and the hook into client-side routing is significant to mention in addition to the "skip-nav" part.

Demonstrates the usage of [@reach/skip-nav](https://reacttraining.com/reach-ui/skip-nav/)

## Add SkipNavLink and SkipNavContent to your Layout

```javascript:layout.js
import { SkipNavLink, SkipNavContent } from "@reach/skip-nav"
import "@reach/skip-nav/styles.css" //this will show/hide the link on focus

> ;<>

Reviewer 2: I wonder if for clarity for users, it would be easier to see this wrapped in a function called App?

This example will show you how to leverage the [@reach/skip-nav](https://reacttraining.com/reach-ui/skip-nav/) component to implement a skip link and also how to focus that link on navigation to a new page in your `gatsby-browser.js` file.

## Using @reach/skip-nav

### Add SkipNavLink and SkipNavContent to your Layout

```javascript:layout.js
import { SkipNavLink, SkipNavContent } from "@reach/skip-nav"
import "@reach/skip-nav/styles.css" //this will show/hide the link on focus

const Layout = ({ children }) => (
  <>
    <SkipNavLink /> //put the link at the very top of your page
    <Navigation />
    <Header />
    <SkipNavContent /> //the link will go to here, so put it right before main content
    <main>{children}</main>
    <Footer />
  </>
)
export default Layout
```

### Focus your link on page navigation

> 

Reviewer: What do you think about adding some more setup info about the API method? This makes me think it should be a doc that is surfaced on gatsbyjs.org, in addition to an example (since the examples aren't currently floated up anywhere else).
```suggestion
Hooking into Gatsby's onRouteUpdate API method will allow you to focus automatically on a skip link on page change, putting a user in a more appropriate spot to take action. More information about this method can be found the [browser API docs](https://www.gatsbyjs.org/docs/browser-apis/#onRouteUpdate).
```

Author: I like that idea! I'm hoping to aggregate future accessibility examples and surface them together anyhow so this can be a good starting point.

    const skipLink = document.querySelector("[data-reach-skip-link]") //this is the query selector that comes with the <SkipNavLink> component
    if (skipLink) {
      skipLink.focus()
    }
  }
}
```

## 🚀 Using this example

### 🧐 What's inside?

A quick look at the relevant files and directories you'll see in this example:

    .
    ├── cypress/
    │   ├── integration/
    │   │   ├── skip-nav.test.js
    ├── src/
    │   ├── components/
    │   │   ├── header.css
    │   │   ├── header.js
    │   │   ├── layout.js
    │   │   └── seo.js
    │   └── pages/
    ├── gatsby-browser.js

1.  **`/src`**: This directory will contain all of the code related to what you will see on the front-end of your site. It has pages and components to be used in those pages.
1.  **`/components`**: This directory will contain all of the code related to what you will see on the front-end of your site. It has pages and components to be used in those pages.

    1.  **`/header.css`**: This ensures that our navigation links have focus rings, which are important visual cues when navigating by keyboard.
    1.  **`/header.js`**: This has the navigation links for your pages.

>     1.  **`/layout.js`**: This component is used in all of our pages to make sure they all use the same navigation. This is also where we are implementing our skip link.

Reviewer: ```suggestion
    1.  **`/layout.js`**: This component is used in all of your pages to make sure they all use the same navigation. This is also where you will implement your skip link.
```

    1.  **`/seo.js`**: This information is necessary to give the site a 100% in [Lighthouse](https://developers.google.com/web/tools/lighthouse) for accessibility since screen readers need a title and language.

> 1.  **`/pages`**: This directory contains pages that will be automatically built and served by Gatsby. We've included three pages to demonstrate navigation between them and how our skip nav link behaves. All of these pages use the `Layout` component.

Reviewer: ```suggestion
1.  **`/pages`**: This directory contains pages that will be automatically built and served by Gatsby. This example includes three pages to demonstrate navigation between them and how your skip nav link behaves. All of these pages use the `Layout` component.
```

1.  **`gatsby-browser.js`**: This file is where we tell the Gatsby to focus the skip navigation when we navigate to a new page.

> 1.  **`/cypress**: This directory is where tests and [Cypress](https://www.cypress.io/) configuration live. We're going to focus on the test. If you want to learn more about using Cypress, check out our [example](https://github.com/gatsbyjs/gatsby/tree/master/examples/using-cypress).

Reviewer: ```suggestion
1.  **`/cypress**: This directory is where tests and [Cypress](https://www.cypress.io/) configuration live. You're going to focus on the test. If you want to learn more about using Cypress, check out the [example](https://github.com/gatsbyjs/gatsby/tree/master/examples/using-cypress).
```
What does "we're going to focus on the test" mean? (I wasn't sure when swapping out pronouns)

> 1.  **`/integrations/skip-nav.test.js`**: runs two tests to ensure that we have a skip link and that the skip link is focused on page navigation.

Reviewer: ```suggestion
1.  **`/integrations/skip-nav.test.js`**: runs two tests to ensure that you have a skip link and that the skip link is focused on page navigation.
```

### Running the example

1. Use the CLI to create run this example

   ```shell
   gatsby develop
   ```

1. Your site is now running at `http://localhost:8000`!
1. Press `tab` and you'll see the skip link!
1. If you press `enter` and then `tab` again you'll see that you skipped over the rest of the navigation and are in the main content!

> ### Runnings tests

Reviewer: ```suggestion
### Running tests
```

Author: ooh Typo bot caught something! oddly satisfying!
